### PR TITLE
Template v2 comments

### PR DESCRIPTION
**Pull request by mikewoodward94**
Based on chat logs of review from @mikaelsimard5 

Actions:
1) Update workdir of Dockerfile
2) Add requirements.txt
3) Fill in default values in config
4) Add num_workers to config
5) Add seed everything
6) Update to classification_report
7) Look in to self._attribute
8) Update to on_validation...
9) Check mixup variables in config
10) Update metrics to how scaphx does it
11) Refine evaluate_best_model
12) Update pretrained to Boolean
13) Add trials to config? (and other tune based things)?
14) pretrained and loss read by getboolean
15) early stopping config file option for patience
16) improve (or get claude to improve) the docstrings for transforms

### CODE REVIEW

File: project/classifier_2d/mlops/train.py [hunk 1]
```diff
@@ -44,41 +45,51 @@ def train(config):
...
+            max_lr = float(config['params']['max_lr']),
+            batch_size = int(config['params']['batch_size']),
             dropout = float(config['params']['dropout']),
-            weighted_loss = config['params']['weighted_loss'],
             train_class_weights = train_class_weights,
             validation_class_weights = validation_class_weights,
-            learning_rate = float(config['params']['lr']),
-            max_lr = float(config['params']['max_lr']),
-            nw_batch_size = int(config['params']['nw_batch_size']),
+            weighted_loss = config['params']['weighted_loss'],
```
Comment (mikaelsimard5) [inline project/classifier_2d/mlops/train.py]: should call getboolean()

File: project/classifier_2d/config/config.cfg [hunk 2]
```diff
@@ -1,34 +1,56 @@
...
+; Mixup Parameters
+MIXUP_ALPHA = 0.2
+CUTMIX_ALPHA = 0.2
+MIXUP_PROB = 0.5
+MIXUP_SWITCH_PROB = 0.5
+MIXUP_MODE = batch
+LABEL_SMOOTHING = 0.1
+
+; Data Parameters
+TEST_FRACTION = 0.2
```
Comment (mikaelsimard5) [inline project/classifier_2d/config/config.cfg]: Wondering if validation_fraction is not more appropriate? As in we're not doing a proper train/val/test split but just train/val, would be more representative.

File: project/classifier_2d/config/config.cfg [hunk 3]
```diff
@@ -1,34 +1,56 @@
...
-ARTIFACT_PATH = s3://mlflow
-
 [project]
 NAME = example_mlflow_project
-;VOLUME_MOUNT = < data path>:/DATA
+LABEL_DICT = {"NEGATIVE": 0, "POSITIVE": 1}
 
 [system]
-USE_GPU = True
+USE_GPU = true
```
Comment (mikaelsimard5) [inline project/classifier_2d/config/config.cfg:6]: I can't find any references to USE_GPU - consider removing from both config files or setting that up as intended?

File: project/classifier_2d/mlops/train.py [hunk 4]
```diff
@@ -44,41 +45,51 @@ def train(config):
...
         mlflow.log_param('n_classes', n_classes)
 
         train_class_weights = dm.data_manifest["train"]["class_weights"]
         validation_class_weights = dm.data_manifest["validation"]["class_weights"]
 
         net = Network(
+            n_classes = n_classes,
+            label_dict = label_dict,
             model_name = config['params']['model'],
             pretrained = config['params']['pretrained'],
```
Comment (mikaelsimard5) [inline project/classifier_2d/mlops/train.py]: should call getboolean()

File: project/classifier_2d/mlops/train.py (hunk 1 above)
Comment (mikaelsimard5) [inline project/classifier_2d/mlops/train.py]: should call getboolean()

File: project/classifier_2d/mlops/train.py [hunk 5]
```diff
@@ -44,41 +45,51 @@ def train(config):
...
-            max_lr = float(config['params']['max_lr']),
-            nw_batch_size = int(config['params']['nw_batch_size']),
+            weighted_loss = config['params']['weighted_loss'],
+            loss_fcn = config['params']['loss_fcn'],
+            weight_decay = float(config['params']['weight_decay']),
+            mixup_alpha = float(config['params']['mixup_alpha']),
+            cutmix_alpha = float(config['params']['cutmix_alpha']),
+            mixup_prob = float(config['params']['mixup_prob']),
+            mixup_switch_prob = float(config['params']['mixup_switch_prob']),
+            mixup_mode = config['params']['mixup_mode'],
```
Comment (mikaelsimard5) [inline project/classifier_2d/mlops/train.py:92]: That's a lot of arguments passed to Network I think. Could we refactor to something like

-> add new field to config file called [mixup], inside of it have the params passed to mixup defined directly ie alpha, cutmix_alpha, prob, ...
-> read config["mixup"] as dict, pass as mixup_kwargs to Network
-> then read self.mixup_fn as Mixup(num_classes=self.n_classes, **mixup_kwargs) or something simpler

In theory it sounds nicer, in practice you have to cast every key inside [mixup] to the correct type so I'm not sure. Works well if we move config file to yaml because you don't need to worry about that :)

Comment (mikaelsimard5) [reply project/classifier_2d/mlops/train.py:92]: We've had a chat about this and decided no kwargs for the moment for the sake of readability (easier to know what's happening) but just thought I'd leave the comment here so we keep track of our discussions.

File: project/classifier_2d/mlops/tune.py [hunk 6]
```diff
@@ -19,14 +21,13 @@
...
-    max_lr = trial.suggest_categorical("max_lr",[5e-4,1e-3])
+    lr = trial.suggest_float("lr", 1e-5, 1e-3, log=True)
+    max_lr = trial.suggest_categorical("max_lr",[1e-4,5e-4,1e-3,5e-3])
     model = trial.suggest_categorical("model", ["convnextv2_tiny.fcmae_ft_in22k_in1k","convnextv2_base.fcmae_ft_in22k_in1k"])
-    batch_size = trial.suggest_int("batch_size", 4, 16, step=4)
-    pretrained = trial.suggest_categorical("pretrained", ["TRUE"])
-    label_smoothing = trial.suggest_float("label_smoothing", 0.0, 0.1)
+    batch_size = trial.suggest_categorical("batch_size", [16, 32, 64])
+    pretrained = trial.suggest_categorical("pretrained", [True, False])
+    label_smoothing = trial.categorical("label_smoothing", [0.0, 0.5, 0.1])
```
Comment (mikaelsimard5) [inline project/classifier_2d/mlops/tune.py]: probably meant 0.05 rather than 0.5?

File: project/classifier_2d/config/config.cfg [hunk 7]
```diff
@@ -1,34 +1,56 @@
...
-PRETRAINED = 
-WEIGHTED_LOSS = 
-LABEL_SMOOTHING = 
-MAX_EPOCHS = 
-GRAD_BATCHES = 
-IMAGE_SIZE = 
+; Model
+MODEL = convnextv2_base.fcmae_ft_in22k_in1k
+PRETRAINED = true
+IMAGE_SIZE = 224
```
Comment (mikaelsimard5) [inline project/classifier_2d/config/config.cfg:27]: Should the default config in the template provide some information on what some variables are for? 

For example IMAGE_SIZE resizes N-dimensional data from [H,W,N] to [IMAGE_SIZE,IMAGE_SIZE,N] before training or so. I may be overthinking this as most variables are self-explanatory :D

File: project/classifier_2d/src/transforms/__init__.py [hunk 8]
```diff
@@ -42,7 +50,21 @@ def load_xnat(xnat_configuration: dict):
...
+    and scales intensity to [0, 255] and casts to float32.
+
+    Args:
+        image_size (int): Target spatial size for the longest side of the image.
+            The image is resized to image_size + 20 here to leave a small margin
+            before the final crop applied by downstream transforms.
+
+    Returns:
+        list[MapTransform]: Ordered list of MONAI transforms operating on the
+            'image' key. Output tensor has shape (1, H, W) with dtype float32
```
Comment (mikaelsimard5) [inline project/classifier_2d/src/transforms/__init__.py]: This could be clearer exactly what are the expected H,W - the output tensor has shape (1, H, W) where max*H, W) = image_size + 20, and the shorter dimension is scaled proportionally to maintain the aspect ratio. 

I'd update the output to include that info and rephrase a bit more the top bit which is incorrect (not exact that the image is resized to image_size + 20 - it's more the larger size is resized to that)

File: project/classifier_2d/src/network.py [hunk 9]
```diff
@@ -163,61 +246,204 @@ def evaluate_best_model(self, model, threshold_tune = False, epoch=0):
...
+        ax.plot(recall, precision, color='darkorange', lw=2, label=f'PR curve (AP = {ap:.3f})')
+        baseline = labels.mean()
+        ax.axhline(y=baseline, color='grey', linestyle='--', lw=1, label=f'Baseline (prevalence = {baseline:.2f})')
+        ax.set_xlabel('Recall (Sensitivity)')
+        ax.set_ylabel('Precision')
+        ax.set_title('Precision-Recall Curve')
+        ax.legend(loc='upper right')
+        plt.tight_layout()
+        mlflow.log_figure(fig, 'evaluation/pr_curve.png')
+        mlflow.log_metric('best_model_avg_precision', ap)
```
Comment (mikaelsimard5) [inline project/classifier_2d/src/network.py]: that's on me from scaphX but just realised this should be called best_model_auprc. 

It's technically equivalent metrics (the auprc or the average precision) but it's less confusing to use auprc I think.

**Review by mikaelsimard5 — CHANGES_REQUESTED** [review]
Pretty good! Just commented on minor issues really, and as we've talked we should add as a feature request to change config towards yaml to make things easier in the future.

File: project/classifier_2d/config/config.cfg (hunk 2 above)
Comment (mikewoodward94) [reply project/classifier_2d/config/config.cfg]: Agreed! I'll also add an issue to add a test split and the resulting inference loop.

File: project/classifier_2d/config/config.cfg (hunk 3 above)
Comment (mikewoodward94) [reply project/classifier_2d/config/config.cfg:6]: This is used by the CSC-MLOps python package! Reads from the config to determine if GPUs should be requested.

File: project/classifier_2d/config/config.cfg (hunk 7 above)
Comment (mikewoodward94) [reply project/classifier_2d/config/config.cfg:27]: I think we can leave as is for now, but if we get repeated comments from users then we can revisit!

File: project/classifier_2d/src/transforms/__init__.py (hunk 8 above)
Comment (mikewoodward94) [reply project/classifier_2d/src/transforms/__init__.py]: I'll blame claude for that one...

**Review by mikewoodward94 — COMMENTED** [review]
Changes made as requested!

**Review by mikaelsimard5 — APPROVED** [review]
All good :) ready to merge!
